nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+
+
+====getStorage====
+Reads the raw value in smart contract storage.
+
+
+'''Method Interface'''
+
+<pre>
+function getStorage({ scriptHash: string, key: string, network?: string }): Promise<StorageResponse>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface GetStorageArgs {
+  scriptHash: string; // script hash of the smart contract to invoke a read on
+  key: string; // key of the storage value to retrieve from the contract
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface StorageResponse {
+  result: string; // The raw value that's stored in the contract
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`RPC_ERROR`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+getStorage({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  key: 'game.status'
+})
+.then(res => {
+  const value = res.result;
+  console.log('Storage value: ' + value);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_REFUSED:
+      console.log('Connection dApp not connected. Please call the "connect" function.');
+      break;
+    case RPC_ERROR:
+      console.log('There was an error when broadcasting this transaction to the network.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hash is valid
+* Broadcast the RPC request
+* Return the RPC response results
+
+Example
+<pre>
+{
+  result: 'hello world'
+}
+</pre>
+
+
+====invokeRead====
+Execute a contract invocation in read-only mode.
+
+
+'''Method Interface'''
+
+<pre>
+function invokeRead({
+  scriptHash: string,
+  operation: string,
+  args: Argument[]
+  network?: string,
+ }): Promise<result: Object>
+</pre>
+
+'''Input arguments'''
+
+<pre>
+interface InvokeReadArgs {
+  scriptHash: string; // script hash of the smart contract to invoke a read on
+  operation: string; // operation on the smart contract to call
+  args: Argument[]; // any input arguments for the operation
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+
+interface Argument {
+  type: ArgumentDataType;
+  value: any;
+}
+
+type ArgumentDataType = 'String'|'Boolean'|'Hash160'|'Hash256'|'Integer'|'ByteArray'|'Array'|'Address';
+</pre>
+
+
+'''Success return value'''
+
+result: Object - Response from RPC node
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`RPC_ERROR`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+invokeRead({
```

Comment:
invokeRead or invokeReadOnly?

Reply:
Yes, this `invokeRead` method is for reading. Whereas `invoke` is for "writing"